accepted/future-releases/class-modifiers/feature-specification.md

```diff
@@ -530,62 +577,282 @@ mixinModifier     ::= 'sealed' | 'base' | 'interface' | 'final'
-    or no `extends` clause. It cannot have any `with` clause.*
+*   Mix in a mixin or mixin class marked `interface`, `final` or `sealed`
+    outside of the library where it is declared.
+
+    ```dart
+    // a.dart
+    interface mixin class I {}
+    final mixin class F {}
+    sealed mixin class S {}
+
+    interface mixin IM {}
+    final mixin FM {}
+    sealed mixin SM {}
+
+    // b.dart
+    import 'a.dart';
+
+    class C1 with I {} // Error.
+    class C2 with F {} // Error.
+    class C3 with S {} // Error.
+
+    class C1 with IM {} // Error.
+    class C2 with FM {} // Error.
+    class C3 with SM {} // Error.
+    ```
+
+A typedef can't be used to subvert these restrictions or any of the restrictions
+below. When extending, implementing, or mixing in a typedef, we look at the
+library where class or mixin the typedef resolves to is defined to determine if
+the behavior is allowed. *Note that the library where the _typedef_ is defined
+does not come into play. Typedefs cannot be marked with any of the new
+modifiers.*
+
+### Disallowing implementation
+
+It is a compile-time error if a subtype of a declaration is marked `base` or
+`final` is not marked `base`, `final`, or `sealed`. This restriction applies to
+both direct and indirect subtypes and along all paths that introduce subtypes:
+`implements` clauses, `extends` clauses, `with` clauses, and `on` clauses. This
+restriction applies even to types within the same library.
+
+*Once the ability to use as an interface is removed, it cannot be reintroduced
+in a subtype. If a class is marked `base` or `final`, you may still implement
+the class's interface inside the same library, but the implementing class must
+again be marked `base`, `final`, or `sealed` to avoid it exposing an
+implementable interface.*
+
+Further, while you can ignore some restrictions on declarations within the same
+library, you can't use that to ignore restrictions inherited from other
+libraries.
+
+We say a class or mixin declaration `D` *can't be implemented locally* if it
+extends, mixes in, implements, or has as on type any declaration `S` where:
+
+*   `S` is from another library than `D`, and `S` has the modifier `base`,
+    `final` or `sealed`, or
+
+    ```dart
+    // a.dart
+    base mixin class S {}
+
+    // b.dart
+    import 'a.dart';
+
+    // These can't be implemented locally:
+    class DE extends S {}
+    class DM with S {}
+    mixin MO on S {}
+    ```
 
-    A _trivial generative constructor_ is a non-redirecting generative constructor 
-    which has
+*   `S` is from the same library as `D`, and `S` can't be implemented locally.
 
-    *   an empty parameter list,
-    *   no initializer list (no `: ...`),  
-    *   no constructor body (only `;`),
-    *   and is not marker `external`.
+    ```dart
+    // a.dart
+    base class B {}
+
+    // b.dart
+    import 'a.dart';
+
+    // This can't be implemented locally (from the previous rule):
+    base mixin class S extends B {}
```

Comment:
Can a mixin class extend `B`?

Reply:
But we just took `on` away from all these constraints, so we just can't create the mixin case.